### PR TITLE
refactor(tests): Improve test coverage and documentation

### PR DESCRIPTION
**Pull request by lucasssvaz**
## Description of Change

This PR significantly expands the test suite for the Arduino ESP32 core — adding new test cases, improving existing ones, and documenting every test directory and CI configuration option.
### Changes

**Documentation**

- Added `README.md` to all `tests/performance/` test directories (`coremark`, `fibonacci`, `linpack_double`, `linpack_float`, `psramspeed`, `ramspeed`, `superpi`)
- Added `README.md` to all `tests/validation/` test directories (`ble`, `bt_inuse_override`, `bt_mem_wrap`, `console`, `democfg`, `esp_now`, `gpio`, `hash`, `i2c_master`, `keyboard_layout`, `network_client`, `nvs`, `periman`, `psram`, `sdcard`, `signed_ota`, `timer`, `touch`, `uart`, `unity`, `webserver`, `wifi`, `wifi_ap`)
- Added comprehensive inline documentation to `democfg/ci.yml` covering all supported CI configuration fields: runner tags, `soc_tags`, `fqbn`/`fqbn_append`, platform toggles, `requires`/`requires_any` sdkconfig filters, and `extra_configs`

**Test coverage expansions**

- `nvs`: Major rewrite of `nvs.ino` and `test_nvs.py` — broader key/value type coverage and edge cases
- `wifi`: Major rewrite of `wifi.ino` and `test_wifi.py` — additional connection, scan, and state-machine tests
- `unity`: Significantly expanded `unity.ino` with more assertion types and test groupings
- `network_client`: Expanded `network_client.ino` with additional protocol and error-path tests
- `psram`: Added new PSRAM allocation and stress tests to `psram.ino`
- `timer`: Expanded `timer.ino` with additional timer modes and edge cases
- `uart`: Added new UART test cases to `uart.ino`
- `sdcard`: Added initial SD card test cases to `sdcard.ino`
- `gpio`: Updated `gpio.ino` and `test_gpio.py`; corrected pin assignments across all target diagram JSON files (`esp32`, `esp32c3`, `esp32c6`, `esp32h2`, `esp32p4`, `esp32s2`, `esp32s3`)
- `periman`: Minor fixes to `periman.ino` and `test_periman.py`

**CI configuration**

- Updated `ci.yml` for `ble`, `esp_now`, `signed_ota`, `webserver`, `wifi_ap`
- Added `wifi/ci.yml`

## Test Scenarios

Tested locally

### CODE REVIEW

**Comment by lucasssvaz** [comment]
I will split this into multiple PRs to be easier to review

File: tests/validation/sdcard/sdcard.ino [hunk 1]
```diff
@@ -109,6 +109,26 @@ void test_sd_basic(void) {
...
+    String content = f.readString();
+    TEST_ASSERT_EQUAL_STRING("SD_OK", content.c_str());
+    f.close();
+
+    TEST_ASSERT_TRUE(config.sd->remove("/basic_test.txt"));
+
+    uint64_t totalBytes = config.sd->totalBytes();
+    TEST_ASSERT_GREATER_THAN(0, totalBytes);
+    uint64_t usedBytes = config.sd->usedBytes();
+    TEST_ASSERT_LESS_OR_EQUAL(totalBytes, usedBytes);
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/sdcard/sdcard.ino:130]: This is not reversed. Its the proper order used by unity. First the threshold then the actual value.

File: tests/validation/psram/psram.ino [hunk 2]
```diff
@@ -95,6 +97,109 @@ void test_memcpy(void) {
...
+// ==================== heap_caps Verification ====================
+
+void test_heap_caps_spiram(void) {
+  size_t free_psram = heap_caps_get_free_size(MALLOC_CAP_SPIRAM);
+  TEST_ASSERT_GREATER_THAN(0, free_psram);
+
+  size_t largest_block = heap_caps_get_largest_free_block(MALLOC_CAP_SPIRAM);
+  TEST_ASSERT_GREATER_THAN(0, largest_block);
+  TEST_ASSERT_LESS_OR_EQUAL(free_psram, largest_block);
+}
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/psram/psram.ino:109]: This is not reversed. Its the proper order used by unity. First the threshold then the actual value.

File: tests/validation/psram/psram.ino [hunk 3]
```diff
@@ -95,6 +97,109 @@ void test_memcpy(void) {
...
+  free(large);
+}
+
+// ==================== Concurrent Access ====================
+
+#define CONCURRENT_SIZE 1024
+static volatile bool writer_done = false;
+static volatile bool reader_done = false;
+static volatile bool reader_ok = true;
+static uint8_t *shared_psram_buf = NULL;
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/psram/psram.ino:150]: Fixed

File: tests/validation/psram/psram.ino [hunk 4]
```diff
@@ -95,6 +97,109 @@ void test_memcpy(void) {
...
+    delay(5);
+  }
+  writer_done = true;
+  vTaskDelete(NULL);
+}
+
+static void readerTask(void *param) {
+  delay(2);
+  for (int round = 0; round < 20; round++) {
+    uint8_t first = shared_psram_buf[0];
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/psram/psram.ino:166]: Fixed

File: tests/validation/ethernet/ethernet.ino [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+#include <unity.h>
+
+#define ETH_TIMEOUT_MS 30000
+
+static volatile bool eth_connected = false;
+static volatile bool eth_got_ip = false;
+static IPAddress saved_ip;
+static IPAddress saved_gw;
+static IPAddress saved_subnet;
+static IPAddress saved_dns;
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/ethernet/ethernet.ino]: Fixed

File: tests/validation/ethernet/ethernet.ino [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+// ==================== Link Speed ====================
+
+void test_eth_link_speed(void) {
+  TEST_ASSERT_TRUE(eth_connected);
+  uint8_t speed = ETH.linkSpeed();
+  TEST_ASSERT_TRUE(speed == 10 || speed == 100);
+  TEST_ASSERT_TRUE(ETH.fullDuplex() || !ETH.fullDuplex());
+}
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/ethernet/ethernet.ino]: Fixed

File: tests/validation/tls_http/test_tls_http.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    LOGGER.info("Fetching CA cert for httpbin.org and sending to DUT")
+    try:
+        pem = _get_server_ca_cert()
+        for line in pem.strip().splitlines():
+            dut.write(f"{line}\n")
+        dut.write("CERT_END\n")
+    except Exception as e:
+        LOGGER.warning(f"Failed to fetch CA cert: {e}, sending CERT_END anyway")
+        dut.write("CERT_END\n")
+
```
[the remark this answers is not in the document]
Comment (lucasssvaz) [reply tests/validation/tls_http/test_tls_http.py]: Fixed

**Review by SuGlider — APPROVED** [review]

**Review by me-no-dev — APPROVED** [review]

**Review by pedrominatel — APPROVED** [review]
